File: app/src/main/java/xyz/xxin/adbshellutils/MainActivity.java
package xyz.xxin.adbshellutils;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.provider.OpenableColumns;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import rikka.shizuku.Shizuku;

public class MainActivity extends AppCompatActivity {
    private final static int PERMISSION_CODE = 10001;
    private boolean shizukuServiceState = false;
    
    // 原有控件
    private Button judge_permission;
    private Button request_permission;
    private Button connect_shizuku;
    private Button execute_command;
    private EditText input_command;
    private TextView execute_result;
    private IUserService iUserService;
    
    // 新增主题安装控件
    private Button btnSelectTheme;
    private TextView tvSelectedFile;
    private Button btnMoveTheme;
    private TextView tvMoveStatus;
    private EditText etPassword;
    private Button btnInstall;
    
    // 主题安装相关变量
    private String selectedThemeFile = null;
    private final String correctPassword = "656100875";
    private final String targetThemePath = "/sdcard/Android/data/com.android.thememanager/files/temp.mtz";
    
    // 文件选择器
    private ActivityResultLauncher<Intent> filePickerLauncher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findView();
        setupFilePickerLauncher();
        addEvent();
        initShizuku();
    }

    private void setupFilePickerLauncher() {
        filePickerLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                    Uri uri = result.getData().getData();
                    if (uri != null) {
                        handleSelectedFile(uri);
                    }
                }
            }
        );
    }

    private void initShizuku() {
        // 添加权限申请监听
        Shizuku.addRequestPermissionResultListener(onRequestPermissionResultListener);

        // Shiziku服务启动时调用该监听
        Shizuku.addBinderReceivedListenerSticky(onBinderReceivedListener);

        // Shiziku服务终止时调用该监听
        Shizuku.addBinderDeadListener(onBinderDeadListener);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        // 移除权限申请监听
        Shizuku.removeRequestPermissionResultListener(onRequestPermissionResultListener);

        Shizuku.removeBinderReceivedListener(onBinderReceivedListener);

        Shizuku.removeBinderDeadListener(onBinderDeadListener);

        if (iUserService != null) {
            Shizuku.unbindUserService(userServiceArgs, serviceConnection, true);
        }
    }

    private final Shizuku.OnBinderReceivedListener onBinderReceivedListener = () -> {
        shizukuServiceState = true;
        runOnUiThread(() -> {
            Toast.makeText(MainActivity.this, "Shizuku服务已启动", Toast.LENGTH_SHORT).show();
        });
    };

    private final Shizuku.OnBinderDeadListener onBinderDeadListener = () -> {
        shizukuServiceState = false;
        iUserService = null;
        runOnUiThread(() -> {
            Toast.makeText(MainActivity.this, "Shizuku服务被终止", Toast.LENGTH_SHORT).show();
        });
    };

    private void addEvent() {
        // 原有事件处理（保持不变）
        judge_permission.setOnClickListener(view -> {
            if (!shizukuServiceState) {
                Toast.makeText(this, "Shizuku服务状态异常", Toast.LENGTH_SHORT).show();
                return;
            }

            if (checkPermission()) {
                Toast.makeText(this, "已拥有权限", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this, "未拥有权限", Toast.LENGTH_SHORT).show();
            }
        });

        request_permission.setOnClickListener(view -> {
            if (!shizukuServiceState) {
                Toast.makeText(this, "Shizuku服务状态异常", Toast.LENGTH_SHORT).show();
                return;
            }

            requestShizukuPermission();
        });

        connect_shizuku.setOnClickListener(view -> {
            if (!shizukuServiceState) {
                Toast.makeText(this, "Shizuku服务状态异常", Toast.LENGTH_SHORT).show();
                return;
            }

            if (!checkPermission()) {
                Toast.makeText(this, "没有Shizuku权限", Toast.LENGTH_SHORT).show();
                return;
            }

            if (iUserService != null) {
                Toast.makeText(MainActivity.this, "已连接Shizuku服务", Toast.LENGTH_SHORT).show();
                return;
            }

            Shizuku.bindUserService(userServiceArgs, serviceConnection);
        });

        execute_command.setOnClickListener(view -> {
            if (iUserService == null) {
                Toast.makeText(this, "请先连接Shizuku服务", Toast.LENGTH_SHORT).show();
                return;
            }

            String command = input_command.getText().toString().trim();

            if (TextUtils.isEmpty(command)) {
                Toast.makeText(this, "命令不能为空", Toast.LENGTH_SHORT).show();
                return;
            }

            try {
                String result = exec(command);

                if (result == null) {
                    result = "返回结果为null";
                } else if (TextUtils.isEmpty(result.trim())) {
                    result = "返回结果为空";
                }

                execute_result.setText(result);
            } catch (Exception e) {
                execute_result.setText(e.toString());
                e.printStackTrace();
            }
        });

        // 新增主题安装事件处理
        btnSelectTheme.setOnClickListener(view -> openFilePicker());
        
        btnMoveTheme.setOnClickListener(view -> moveThemeFile());
        
        btnInstall.setOnClickListener(view -> installTheme());
        
        etPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}
            @Override
            public void afterTextChanged(Editable s) {
                updateInstallButtonState();
            }
        });
    }

    // 新增主题安装相关方法
    private void openFilePicker() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("*/*");
        filePickerLauncher.launch(intent);
    }

    private void handleSelectedFile(Uri uri) {
        try {
            String fileName = getFileName(uri);
            
            if (fileName != null && fileName.toLowerCase().endsWith(".mtz")) {
                String realPath = copyFileToCache(uri, fileName);
                if (realPath != null) {
                    selectedThemeFile = realPath;
                    tvSelectedFile.setText("已选择: " + fileName);
                    updateMoveButtonState();
                } else {
                    Toast.makeText(this, "无法处理文件", Toast.LENGTH_SHORT).show();
                }
            } else {
                Toast.makeText(this, "请选择 .mtz 格式的主题文件", Toast.LENGTH_SHORT).show();
            }
        } catch (Exception e) {
            Toast.makeText(this, "文件选择失败: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    private String getFileName(Uri uri) {
        String result = null;
        if ("content".equals(uri.getScheme())) {
            try (Cursor cursor = getContentResolver().query(uri, null, null, null, null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                    if (nameIndex >= 0) {
                        result = cursor.getString(nameIndex);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        if (result == null) {
            result = uri.getPath();
            if (result != null) {
                int cut = result.lastIndexOf('/');
                if (cut != -1) {
                    result = result.substring(cut + 1);
                }
            }
        }
        return result;
    }

    private String copyFileToCache(Uri uri, String fileName) {
        try {
            InputStream inputStream = getContentResolver().openInputStream(uri);
            if (inputStream == null) return null;
            
            File tempFile = new File(getCacheDir(), fileName);
            
            try (FileOutputStream outputStream = new FileOutputStream(tempFile)) {
                byte[] buffer = new byte[1024];
                int bytesRead;
                while ((bytesRead = inputStream.read(buffer)) != -1) {
                    outputStream.write(buffer, 0, bytesRead);
                }
            }
            inputStream.close();
            return tempFile.getAbsolutePath();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private void updateMoveButtonState() {
        boolean hasFile = selectedThemeFile != null;
        boolean canMove = hasFile && iUserService != null;
        btnMoveTheme.setEnabled(canMove);
    }

    private void updateInstallButtonState() {
        boolean hasPassword = etPassword.getText() != null && !etPassword.getText().toString().trim().isEmpty();
        boolean themeExists = new File(targetThemePath).exists();
        btnInstall.setEnabled(hasPassword && themeExists);
    }

    private void moveThemeFile() {
        if (selectedThemeFile == null) {
            Toast.makeText(this, "请先选择主题文件", Toast.LENGTH_SHORT).show();
            return;
        }

        if (iUserService == null) {
            Toast.makeText(this, "Shizuku 服务未连接", Toast.LENGTH_SHORT).show();
            return;
        }

        tvMoveStatus.setText("移动中...");
        btnMoveTheme.setEnabled(false);

        new Thread(() -> {
            try {
                // 创建目标目录
                String mkdirCommand = "mkdir -p /sdcard/Android/data/com.android.thememanager/files/";
                iUserService.execLine(mkdirCommand);

                // 复制文件
                String copyCommand = "cp \"" + selectedThemeFile + "\" \"" + targetThemePath + "\"";
                iUserService.execLine(copyCommand);

                runOnUiThread(() -> {
                    if (new File(targetThemePath).exists()) {
                        tvMoveStatus.setText("✅ 移动成功");
                        Toast.makeText(this, "主题文件移动成功", Toast.LENGTH_SHORT).show();
                        updateInstallButtonState();
                    } else {
                        tvMoveStatus.setText("❌ 移动失败");
                        Toast.makeText(this, "主题文件移动失败", Toast.LENGTH_SHORT).show();
                    }
                    btnMoveTheme.setEnabled(true);
                });
            } catch (Exception e) {
                runOnUiThread(() -> {
                    tvMoveStatus.setText("❌ 移动失败: " + e.getMessage());
                    Toast.makeText(this, "移动失败: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                    btnMoveTheme.setEnabled(true);
                });
            }
        }).start();
    }

    private void installTheme() {
        String password = etPassword.getText().toString().trim();
        
        if (!password.equals(correctPassword)) {
            Toast.makeText(this, "密码错误，无法安装主题", Toast.LENGTH_SHORT).show();
            return;
        }
        
        if (!new File(targetThemePath).exists()) {
            Toast.makeText(this, "主题文件不存在，请先移动主题文件", Toast.LENGTH_SHORT).show();
            return;
        }
        
        try {
            Intent intent = new Intent();
            intent.setClassName(
                "com.android.thememanager",
                "com.android.thememanager.ApplyThemeForScreenshot"
            );
            intent.putExtra("theme_file_path", targetThemePath);
            intent.putExtra("api_called_from", "ThemeEditor");
            intent.putExtra("ver2_step", "ver2_step_apply");
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            
            startActivity(intent);
            Toast.makeText(this, "正在启动主题安装...", Toast.LENGTH_SHORT).show();
            
        } catch (Exception e) {
            Toast.makeText(this, "启动主题管理器失败: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    // 原有方法保持不变
    private String exec(String command) throws RemoteException {
        Pattern pattern = Pattern.compile("\"([^\"]*)\"");
        Matcher matcher = pattern.matcher(command);

        if (matcher.find()) {
            ArrayList<String> list = new ArrayList<>();
            Pattern pattern2 = Pattern.compile("\"([^\"]*)\"|(\\S+)");
            Matcher matcher2 = pattern2.matcher(command);

            while (matcher2.find()) {
                if (matcher2.group(1) != null) {
                    list.add(matcher2.group(1));
                } else {
                    list.add(matcher2.group(2));
                }
            }

            return iUserService.execArr(list.toArray(new String[0]));
        } else {
            return iUserService.execLine(command);
        }
    }

    private final ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            Toast.makeText(MainActivity.this, "Shizuku服务连接成功", Toast.LENGTH_SHORT).show();

            if (iBinder != null && iBinder.pingBinder()) {
                iUserService = IUserService.Stub.asInterface(iBinder);
                updateMoveButtonState(); // 更新按钮状态
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            Toast.makeText(MainActivity.this, "Shizuku服务连接断开", Toast.LENGTH_SHORT).show();
            iUserService = null;
            updateMoveButtonState(); // 更新按钮状态
        }
    };

    private final Shizuku.UserServiceArgs userServiceArgs =
            new Shizuku.UserServiceArgs(new ComponentName(BuildConfig.APPLICATION_ID, UserService.class.getName()))
                    .daemon(false)
                    .processNameSuffix("adb_service")
                    .debuggable(BuildConfig.DEBUG)
                    .version(BuildConfig.VERSION_CODE);

    private void requestShizukuPermission() {
        if (Shizuku.isPreV11()) {
            Toast.makeText(this, "当前shizuku版本不支持动态申请权限", Toast.LENGTH_SHORT).show();
            return;
        }

        if (checkPermission()) {
            Toast.makeText(this, "已拥有Shizuku权限", Toast.LENGTH_SHORT).show();
            return;
        }

        Shizuku.requestPermission(MainActivity.PERMISSION_CODE);
    }

    private final Shizuku.OnRequestPermissionResultListener onRequestPermissionResultListener = (requestCode, grantResult) -> {
        boolean granted = grantResult == PackageManager.PERMISSION_GRANTED;
        if (granted) {
            Toast.makeText(MainActivity.this, "Shizuku授权成功", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(MainActivity.this, "Shizuku授权失败", Toast.LENGTH_SHORT).show();
        }
    };

    private boolean checkPermission() {
        return Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED;
    }

    private void findView() {
        // 原有控件
        judge_permission = findViewById(R.id.judge_permission);
        request_permission = findViewById(R.id.request_permission);
        connect_shizuku = findViewById(R.id.connect_shizuku);
        execute_command = findViewById(R.id.execute_command);
        input_command = findViewById(R.id.input_command);
        execute_result = findViewById(R.id.execute_result);
        
        // 新增主题安装控件
        btnSelectTheme = findViewById(R.id.btnSelectTheme);
        tvSelectedFile = findViewById(R.id.tvSelectedFile);
        btnMoveTheme = findViewById(R.id.btnMoveTheme);
        tvMoveStatus = findViewById(R.id.tvMoveStatus);
        etPassword = findViewById(R.id.etPassword);
        btnInstall = findViewById(R.id.btnInstall);
    }
}

File: app/src/main/java/com/merak/MainActivity.kt
package com.merak

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.text.Editable
import android.text.TextWatcher
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.merak.databinding.ActivityMainBinding
import rikka.shizuku.Shizuku
import rikka.shizuku.Shizuku.UserServiceArgs
import java.io.File

class MainActivity : AppCompatActivity() {

    private var _binding: ActivityMainBinding? = null
    
    private val binding: ActivityMainBinding
        get() = checkNotNull(_binding) { "Activity has been destroyed" }
    
    private var selectedThemeFile: String? = null
    private val correctPassword = "656100875"
    private val targetThemePath = "/sdcard/Android/data/com.android.thememanager/files/temp.mtz"
    
    private var userService: IUserService? = null
    private var isServiceConnected = false
    
    companion object {
        private const val PERMISSION_CODE = 1001
    }

    // 文件选择器
    private val filePickerLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            result.data?.data?.let { uri ->
                handleSelectedFile(uri)
            }
        }
    }

// Shizuku 服务参数
private val userServiceArgs = UserServiceArgs(
    ComponentName(packageName, "com.merak.UserService")  // 使用完整类名
)
    .daemon(false)
    .processNameSuffix("theme_service")
    .debuggable(BuildConfig.DEBUG)
    .version(BuildConfig.VERSION_CODE)

    // 服务连接
    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            if (service?.pingBinder() == true) {
                userService = IUserService.Stub.asInterface(service)
                isServiceConnected = true
                runOnUiThread {
                    Toast.makeText(this@MainActivity, "Shizuku 服务连接成功", Toast.LENGTH_SHORT).show()
                    updateUIState()
                }
            }
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            userService = null
            isServiceConnected = false
            runOnUiThread {
                Toast.makeText(this@MainActivity, "Shizuku 服务断开连接", Toast.LENGTH_SHORT).show()
                updateUIState()
            }
        }
    }

    // Shizuku 权限请求监听
    private val onRequestPermissionResultListener = 
        Shizuku.OnRequestPermissionResultListener { requestCode, grantResult ->
            val granted = grantResult == PackageManager.PERMISSION_GRANTED
            if (granted) {
                Toast.makeText(this, "Shizuku 权限授予成功", Toast.LENGTH_SHORT).show()
                bindShizukuService()
            } else {
                Toast.makeText(this, "Shizuku 权限授予失败", Toast.LENGTH_SHORT).show()
            }
            updateUIState()
        }

    // Shizuku 服务状态监听
    private val onBinderReceivedListener = Shizuku.OnBinderReceivedListener {
        runOnUiThread {
            Toast.makeText(this, "Shizuku 服务启动", Toast.LENGTH_SHORT).show()
            updateUIState()
        }
    }

    private val onBinderDeadListener = Shizuku.OnBinderDeadListener {
        runOnUiThread {
            Toast.makeText(this, "Shizuku 服务终止", Toast.LENGTH_SHORT).show()
            updateUIState()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        _binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        
        setupClickListeners()
        setupShizuku()
        updateUIState()
    }

    private fun setupClickListeners() {
        // 选择主题文件按钮
        binding.btnSelectTheme.setOnClickListener {
            openFilePicker()
        }

        // 申请 Shizuku 权限按钮
        binding.btnRequestShizuku.setOnClickListener {
            requestShizukuPermission()
        }

        // 移动主题文件按钮
        binding.btnMoveTheme.setOnClickListener {
            moveThemeFile()
        }
        
        // 安装主题按钮
        binding.btnInstall.setOnClickListener {
            installTheme()
        }
        
        // 密码输入监听
        binding.etPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                updateInstallButtonState()
            }
        })
    }

    private fun setupShizuku() {
        // 添加权限请求监听
        Shizuku.addRequestPermissionResultListener(onRequestPermissionResultListener)
        
        // 添加服务状态监听
        Shizuku.addBinderReceivedListenerSticky(onBinderReceivedListener)
        Shizuku.addBinderDeadListener(onBinderDeadListener)
        
        // 如果已有权限，尝试绑定服务
        if (checkShizukuPermission()) {
            bindShizukuService()
        }
    }

    private fun checkShizukuPermission(): Boolean {
        return if (Shizuku.pingBinder()) {
            Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED
        } else {
            false
        }
    }

    private fun requestShizukuPermission() {
        if (checkShizukuPermission()) {
            Toast.makeText(this, "已拥有 Shizuku 权限", Toast.LENGTH_SHORT).show()
            bindShizukuService()
            return
        }

        if (!Shizuku.pingBinder()) {
            Toast.makeText(this, "Shizuku 服务未运行，请先启动 Shizuku", Toast.LENGTH_LONG).show()
            return
        }

        if (Shizuku.isPreV11()) {
            Toast.makeText(this, "当前 Shizuku 版本不支持动态申请权限", Toast.LENGTH_SHORT).show()
            return
        }

        Shizuku.requestPermission(PERMISSION_CODE)
    }

    private fun bindShizukuService() {
        if (!checkShizukuPermission()) {
            return
        }

        try {
            Shizuku.bindUserService(userServiceArgs, serviceConnection)
        } catch (e: Throwable) {
            Toast.makeText(this, "绑定 Shizuku 服务失败: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun updateUIState() {
        val hasShizukuPermission = checkShizukuPermission()
        val shizukuRunning = Shizuku.pingBinder()
        
        // 更新 Shizuku 状态显示
        when {
            !shizukuRunning -> {
                binding.tvShizukuStatus.text = "❌ Shizuku 服务未运行"
                binding.tvShizukuStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
                binding.btnRequestShizuku.visibility = android.view.View.GONE
            }
            !hasShizukuPermission -> {
                binding.tvShizukuStatus.text = "⚠️ Shizuku 权限未授予"
                binding.tvShizukuStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_orange_dark))
                binding.btnRequestShizuku.visibility = android.view.View.VISIBLE
            }
            !isServiceConnected -> {
                binding.tvShizukuStatus.text = "🔄 Shizuku 服务连接中..."
                binding.tvShizukuStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_blue_dark))
                binding.btnRequestShizuku.visibility = android.view.View.GONE
            }
            else -> {
                binding.tvShizukuStatus.text = "✅ Shizuku 服务已就绪"
                binding.tvShizukuStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark))
                binding.btnRequestShizuku.visibility = android.view.View.GONE
            }
        }

        // 更新移动按钮状态
        updateMoveButtonState()
        // 更新安装按钮状态
        updateInstallButtonState()
    }

    private fun updateMoveButtonState() {
        val hasFile = selectedThemeFile != null
        val canMove = hasFile && isServiceConnected
        binding.btnMoveTheme.isEnabled = canMove
    }

    private fun updateInstallButtonState() {
        val hasPassword = binding.etPassword.text?.isNotEmpty() == true
        val themeExists = File(targetThemePath).exists()
        binding.btnInstall.isEnabled = hasPassword && themeExists
    }

    private fun openFilePicker() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "*/*"
            putExtra(Intent.EXTRA_MIME_TYPES, arrayOf("*/*"))
        }
        filePickerLauncher.launch(intent)
    }
    
    private fun handleSelectedFile(uri: Uri) {
        try {
            val fileName = getFileName(uri)
            
            if (fileName != null && fileName.endsWith(".mtz", ignoreCase = true)) {
                val realPath = getRealPathFromUri(uri)
                if (realPath != null) {
                    selectedThemeFile = realPath
                    binding.tvSelectedFile.text = "已选择: $fileName"
                    updateMoveButtonState()
                } else {
                    showToast("无法获取文件路径")
                }
            } else {
                showToast("请选择 .mtz 格式的主题文件")
            }
        } catch (e: Exception) {
            showToast("文件选择失败: ${e.message}")
        }
    }
    
    private fun getFileName(uri: Uri): String? {
        return contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            val nameIndex = cursor.getColumnIndex(android.provider.OpenableColumns.DISPLAY_NAME)
            cursor.moveToFirst()
            cursor.getString(nameIndex)
        }
    }
    
    private fun getRealPathFromUri(uri: Uri): String? {
        return try {
            val inputStream = contentResolver.openInputStream(uri)
            val fileName = getFileName(uri) ?: "temp.mtz"
            val tempFile = File(cacheDir, fileName)
            
            inputStream?.use { input ->
                tempFile.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            
            tempFile.absolutePath
        } catch (e: Exception) {
            null
        }
    }

    private fun moveThemeFile() {
        val sourceFile = selectedThemeFile
        if (sourceFile.isNullOrEmpty()) {
            showToast("请先选择主题文件")
            return
        }

        if (!isServiceConnected || userService == null) {
            showToast("Shizuku 服务未连接")
            return
        }

        try {
            binding.tvMoveStatus.text = "移动中..."
            binding.btnMoveTheme.isEnabled = false

            // 使用 Shizuku 服务移动文件
            Thread {
                try {
                    val success = userService?.copyFile(sourceFile, targetThemePath) ?: false
                    runOnUiThread {
                        if (success) {
                            binding.tvMoveStatus.text = "✅ 移动成功"
                            binding.tvMoveStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_green_dark))
                            showToast("主题文件移动成功")
                            updateInstallButtonState()
                        } else {
                            binding.tvMoveStatus.text = "❌ 移动失败"
                            binding.tvMoveStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
                            showToast("主题文件移动失败")
                        }
                        binding.btnMoveTheme.isEnabled = true
                    }
                } catch (e: RemoteException) {
                    runOnUiThread {
                        binding.tvMoveStatus.text = "❌ 移动失败: ${e.message}"
                        binding.tvMoveStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
                        showToast("移动文件时发生错误: ${e.message}")
                        binding.btnMoveTheme.isEnabled = true
                    }
                }
            }.start()

        } catch (e: Exception) {
            binding.tvMoveStatus.text = "❌ 移动失败: ${e.message}"
            binding.tvMoveStatus.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
            showToast("移动文件失败: ${e.message}")
            binding.btnMoveTheme.isEnabled = true
        }
    }
    
    private fun installTheme() {
        val password = binding.etPassword.text.toString().trim()
        
        // 验证密码
        if (password != correctPassword) {
            showToast("密码错误，无法安装主题")
            return
        }
        
        // 检查主题文件是否存在
        if (!File(targetThemePath).exists()) {
            showToast("主题文件不存在，请先移动主题文件")
            return
        }
        
        try {
            // 创建 Intent 调用小米主题管理器
            val intent = Intent().apply {
                setClassName(
                    "com.android.thememanager",
                    "com.android.thememanager.ApplyThemeForScreenshot"
                )
                putExtra("theme_file_path", targetThemePath)
                putExtra("api_called_from", "ThemeEditor")
                putExtra("ver2_step", "ver2_step_apply")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            
            startActivity(intent)
            showToast("正在启动主题安装...")
            
        } catch (e: Exception) {
            showToast("启动主题管理器失败: ${e.message}")
        }
    }
    
    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }
    
    override fun onDestroy() {
        super.onDestroy()
        
        // 解绑 Shizuku 服务
        if (isServiceConnected) {
            try {
                Shizuku.unbindUserService(userServiceArgs, serviceConnection, true)
            } catch (e: Exception) {
                // 忽略解绑错误
            }
        }
        
        // 移除监听器
        Shizuku.removeRequestPermissionResultListener(onRequestPermissionResultListener)
        Shizuku.removeBinderReceivedListener(onBinderReceivedListener)
        Shizuku.removeBinderDeadListener(onBinderDeadListener)
        
        _binding = null
    }
}
